magefile.go

```diff
@@ -6,26 +6,188 @@ package main
 import (
 	"fmt"
 	"os"
+	"os/exec"
 	"path"
+	"strings"
+	"time"
 
+	"github.com/go-logr/logr"
 	"github.com/magefile/mage/mg"
 	"github.com/magefile/mage/sh"
 	"github.com/mt-sre/devkube/magedeps"
 )
 
+const (
+	module = "github.com/openshift/addon-operator"
+)
+
 // Directories
 var (
 	// Working directory of the project.
 	workDir string
 	// Dependency directory.
 	depsDir magedeps.DependencyDirectory
+	logger  *logr.Logger
+)
+
+// Building
+// --------
+type Build mg.Namespace
+
+// Build Tags
+var (
+	branch        string
+	shortCommitID string
+	version       string
+	buildDate     string
+
+	ldFlags string
 )
 
+// init build variables
+func (Build) init() error {
+	branchCmd := exec.Command("git", "rev-parse", "--abbrev-ref", "HEAD")
+	branchBytes, err := branchCmd.Output()
+	if err != nil {
+		panic(fmt.Errorf("getting git branch: %w", err))
+	}
+	branch = strings.TrimSpace(string(branchBytes))
+
+	shortCommitIDCmd := exec.Command("git", "rev-parse", "--short", "HEAD")
+	shortCommitIDBytes, err := shortCommitIDCmd.Output()
+	if err != nil {
+		panic(fmt.Errorf("getting git short commit id"))
+	}
+	shortCommitID = strings.TrimSpace(string(shortCommitIDBytes))
+
+	version = strings.TrimSpace(os.Getenv("VERSION"))
+	if len(version) == 0 {
+		version = shortCommitID
+	}
+
+	buildDate = fmt.Sprint(time.Now().UTC().Unix())
+	ldFlags = fmt.Sprintf(`-X %s/internal/version.Version=%s`+
+		`-X %s/internal/version.Branch=%s`+
+		`-X %s/internal/version.Commit=%s`+
+		`-X %s/internal/version.BuildDate=%s`,
+		module, version,
+		module, branch,
+		module, shortCommitID,
+		module, buildDate,
+	)
+	return nil
+}
+
+// Builds binaries from /cmd directory.
+func (Build) cmd(cmd, goos, goarch string) error {
+	mg.Deps(Build.init)
+
+	env := map[string]string{
+		"GOFLAGS":     "",
+		"CGO_ENABLED": "0",
+		"LDFLAGS":     ldFlags,
+	}
+
+	bin := path.Join("bin", cmd)
+	if len(goos) != 0 && len(goarch) != 0 {
+		// change bin path to point to a sudirectory when cross compiling
+		bin = path.Join("bin", goos+"_"+goarch, cmd)
+		env["GOARGS"] = fmt.Sprintf("GOOS=%s GOARCH=%s", goos, goarch)
+	}
+
+	if err := sh.RunWithV(
+		env,
+		"go", "build", "-v", "-o", bin, "./cmd/"+cmd+"/main.go",
+	); err != nil {
+		return fmt.Errorf("compiling cmd/%s: %w", cmd, err)
+	}
```

Comment:
```suggestion
	if err := sh.RunWithV(
		env,
		"go", "build", "-ldflags", fmt.Sprintf("\"-w %s\"", ldFlags), "-v", "-o", bin, "./cmd/"+cmd+"/main.go",
	); err != nil {
		return fmt.Errorf("compiling cmd/%s: %w", cmd, err)
	}
```

Suggesting this because I am not sure if just making ldflags be a part of `env` would ensure `go build` to automatically feed the linker flags from there.

Feel free to correct me here :)